[Jovin Ang] iP

Jovin-Ang opened the pull request:
# Shade chatbot

> Check this out!

Shade is a command-line task management application that helps users track different types of tasks with persistent storage.

## Features

* **Task Types**:
  * 📝 Todo: Simple tasks without dates
  * ⏰ Deadline: Tasks with due dates/times
  * 🗓 Event: Tasks with start/end times
* **Persistence**: Automatic JSON save/load using Gson
* **Command Interface**:
   - Add/delete tasks
   - Mark tasks as complete/incomplete
   - List all tasks with search
   - Interactive help system

Upcoming
- [ ] A GUI!

## Installation

1. Go [here](https://github.com/Jovin-Ang/ip)
1. It uses `gradle`!

```java
public class Main {
    public static void main(String[] args) {
        System.out.println("Lets go!");
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

Jovin-Ang committed:
Refactor Duke to Shade

Jovin-Ang committed:
Update README.md

Jovin-Ang committed:
Implement Greet and Exit

Jovin-Ang committed:
Implement Echo

Jovin-Ang committed:
Implement task add and list

Jovin-Ang committed:
Make task list immutable

Jovin-Ang committed:
Refactor chatbot architecture and add new command system

Jovin-Ang committed:
Removed unnecessary toString() call

Jovin-Ang committed:
Enhance task handling

Added dedicated commands for Todo, Event, and Deadline tasks, along
with corresponding task types (TodoTask, EventTask, DeadlineTask).
Introduced a HelpCommand for listing supported commands.

Jovin-Ang committed:
Prepare test inputs

Jovin-Ang committed:
Added tests

Jovin-Ang committed:
Added delete command

Jovin-Ang committed:
Added tests for delete command

Jovin-Ang committed:
Validate command input for empty and incomplete arguments.

Jovin-Ang committed:
Added tests for empty and incorrect arguments

Jovin-Ang committed:
Make help command dynamic

Add trigger, description and usage fields to Command objects. Add
getCommandUsage() method to Command objects which is used in
HelpCommand.java.

Move user defined exceptions into exception package.

Jovin-Ang committed:
Update help command test case

Jovin-Ang committed:
Refactor package structure and improve command error handling

Jovin-Ang committed:
Undo accidental move of main class

Jovin-Ang committed:
Update tests for incorrect command arguments

Jovin-Ang committed:
Merge branch 'branch-A-MoreOOP'

Jovin-Ang committed:
Merge branch 'add-gradle-support'

Jovin-Ang committed:
Update Gradle configuration and dependencies

Jovin-Ang committed:
Implement JSON storage for task management

Jovin-Ang committed:
Add StorageOperationException for improved error handling

Jovin-Ang committed:
Merge branch 'branch-Level-7'

Jovin-Ang committed:
Refactor task classes to use LocalDateTime for deadlines and event times

Jovin-Ang committed:
Merge branch 'branch-Level-8'

Jovin-Ang committed:
Add unit tests for ToDoTask

Jovin-Ang committed:
Add unit tests for TaskList, DeadlineTask, EventTask

Jovin-Ang committed:
Refactor import statements

Jovin-Ang committed:
Add Checkstyle configuration; refactor code to follow coding standards

Jovin-Ang committed:
feat: Add FindCommand to search for tasks by keyword in TaskList

Jovin-Ang committed:
feat: Add FindCommand to ChatBot

Jovin-Ang committed:
test: Add unit tests for matching task details in TaskList

Jovin-Ang committed:
Merge branch 'branch-A-CodingStandard'

Jovin-Ang committed:
Merge branch 'branch-Level-9'

Jovin-Ang committed:
style: Fix FindCommand and tests to match coding standards

Jovin-Ang committed:
docs: Update README.md

Jovin-Ang committed:
docs: Update formatting in README.md

Jovin-Ang committed:
test: Add HelpCommand tests and IoHandlerStub for command testing

mihirniyogi commented on src/main/java/Shade.java:
```diff
@@ -0,0 +1,27 @@
+import java.util.Scanner;
```
Perhaps every class should placed in a package? To follow the coding standards they provided

Jovin-Ang replied:
Will do in my next commit!

mihirniyogi commented on src/main/java/chatbot/commands/DeadlineCommand.java:
```diff
@@ -0,0 +1,71 @@
+        taskList.addTask(newDeadlineTask);
+        ioHandler.send("Got it. I've added this task:\n  "
+                + newDeadlineTask.getDetails() + "\nYou have "
+                + taskList.getTotalTasks() + " tasks in the list.");
```
i like the clean multi-line wrapped command here, lines 67-69

mihirniyogi commented on src/main/java/chatbot/data/tasks/Task.java:
```diff
@@ -0,0 +1,84 @@
+            throw new IllegalArgumentException("Task cannot be null or empty");
+        }
+        this.task = task;
+        this.isCompleted = false;
```
good use of boolean variable naming (`isComplete`), here and at multiple other places!

mihirniyogi commented on src/main/java/chatbot/ui/ChatBot.java:
```diff
@@ -0,0 +1,167 @@
+    /**
+     * The name of the chatbot.
+     */
+    private final String name;
+    /**
+     * A Scanner object for reading user input.
+     */
+    private final Scanner scanner;
+    /**
+     * A storage manager that handles JSON serialization and deserialization of tasks.
+     */
+    private final JsonStorage storage = new JsonStorage();
+    /**
+     * A map that associates command names with their respective command implementations.
+     */
+    private final Map<String, Command> commands = new LinkedHashMap<>();
+    /**
+     * A TaskList instance, which is used to organize,
+     * store, and manipulate tasks within the chatbot application.
+     */
+    private TaskList taskList;
```
hmm maybe an extra line would be better between logical blocks, so that we know the comment is referring to the variable directly below it, but minor thing
e.g. below lines 39, 43, 47, 51...

mihirniyogi commented on src/main/java/chatbot/commands/QuitCommand.java:
```diff
@@ -0,0 +1,37 @@
+package chatbot.commands;
```
hmm im not so sure about whether package names shld be singular or plural, so i asked a question in the forum ahaha, but minor thing

mihirniyogi commented on src/main/java/chatbot/commands/UnmarkCommand.java:
```diff
@@ -0,0 +1,61 @@
+     */
+    @Override
+    public void execute(String arguments) throws InvalidCommandSyntaxException {
+        if (taskList.getTotalTasks() == 0) {
+            ioHandler.send("No tasks to unmark.");
+            return;
+        }
+        try {
```
maybe an extra line between 48 and 49 for readability?

mihirniyogi commented on src/main/java/chatbot/data/tasks/Task.java:
```diff
@@ -0,0 +1,84 @@
+     *
+     * @throws IllegalTaskStateChangeException if the task is already in the incomplete state.
+     */
+    public void incomplete() throws IllegalTaskStateChangeException {
```
yea 'incomplete' as a verb here is a little less clear than the complete() method above, perhaps `markAsIncomplete()`?

mihirniyogi commented on src/main/java/chatbot/data/tasks/Task.java:
```diff
@@ -0,0 +1,84 @@
+     *
+     * @throws IllegalTaskStateChangeException if the task is already in the completed state.
+     */
+    public void complete() throws IllegalTaskStateChangeException {
```
i would modify this method name to something more 'verb'-y, like `markAsComplete()` but personal preference

mihirniyogi commented on src/main/java/chatbot/data/tasks/Task.java:
```diff
@@ -0,0 +1,84 @@
+    /**
+     * Identifier or label of the task.
+     */
+    private final String task;
```
perhaps 'description' is a more apt variable name... but since it's private it doesn't matter as much. also this is more about code quality than standards, so up to you

mihirniyogi commented on src/main/java/chatbot/storage/TaskTypeAdapter.java:
```diff
@@ -0,0 +1,71 @@
+    /**
+     * The field name used to store the type of the task.
+     */
+    private static final String TYPE_FIELD = "type";
+    /**
+     * The field name used to store the task data.
+     */
+    private static final String DATA_FIELD = "data";
```
nice job on the declaring and naming of constants!

mihirniyogi commented on src/main/java/chatbot/storage/JsonStorage.java:
```diff
@@ -0,0 +1,93 @@
+     */
+    public JsonStorage(String savePath) {
+        this.savePath = Paths.get(savePath);
+        this.gson = new GsonBuilder()
+                .registerTypeAdapter(Task.class, new TaskTypeAdapter())
+                .registerTypeAdapter(LocalDateTime.class, new LocalDateTimeAdapter())
+                .setPrettyPrinting()
+                .create();
+    }
```
great formatting!

mihirniyogi reviewed (commented):
Overall, coding standards were basically almost completely followed, just a few minor stuff, nice work man👍

Jovin-Ang committed:
refactor: Move Shade class to chatbot package

Jovin-Ang committed:
feat: Implement JavaFX GUI for chatbot

Jovin-Ang committed:
Merge branch 'branch-Level-10'

chen-xiangrui commented on src/main/java/chatbot/commands/Command.java:
```diff
@@ -0,0 +1,55 @@
+ *
+ * @author Jovin Ang
+ */
+public abstract class Command {
```
I like the idea of creating abstract classes for Command and observing the polymorphism OOP principle for it.

chen-xiangrui commented on src/main/java/chatbot/data/TaskList.java:
```diff
@@ -0,0 +1,136 @@
+     * @throws IllegalTaskStateChangeException If the task is already in the incomplete state.
+     * @throws IndexOutOfBoundsException       If the specified index is out of bounds in the task list.
+     */
+    public void incompleteTask(int i) throws IllegalTaskStateChangeException, IndexOutOfBoundsException {
```
You could consider using verbs instead of nouns for the method name

chen-xiangrui commented on src/test/java/chatbot/data/tasks/DeadlineTaskTest.java:
```diff
@@ -0,0 +1,79 @@
+package chatbot.data.tasks;
+
+import static org.junit.jupiter.api.Assertions.assertDoesNotThrow;
```
I like the organisation and grouping of the different imported libraries

chen-xiangrui reviewed (commented):
LGTM! The overall structure of the repo is neat, and I can see solid efforts to observe the various OOP principles too.

Jovin-Ang committed:
fix: Validate task inputs in command classes with assertions

Add input validation checks across command classes to prevent invalid
task processing. This ensures data integrity and helps catch potential
issues early in the execution flow.

Jovin-Ang committed:
refactor: Rename task completion methods for clarity

* Renamed methods to better reflect their state-changing nature:
  * complete() -> markAsCompleted()
  * incomplete() -> markAsIncomplete()
  * completeTask() -> markTaskAsCompleted()
  * incompleteTask() -> markTaskAsIncomplete()

* Updated corresponding JUnit test method names to match
  the new method naming convention

This change improves code readability by using verb phrases
that clearly indicate state transitions

Jovin-Ang committed:
refactor: Simplify command execution logic in ChatBot

Jovin-Ang committed:
Merge pull request #1 from Jovin-Ang/branch-A-Assertions

Add assertions to command classes

Jovin-Ang committed:
Merge branch 'master' into branch-A-CodeQuality

Jovin-Ang committed:
Merge pull request #2 from Jovin-Ang/branch-A-CodeQuality

Improve code quality

Jovin-Ang committed:
feat: Add GitHub Actions workflow for Java CI with Gradle

Jovin-Ang committed:
fix: Fix workflow name in GitHub Actions configuration

Jovin-Ang committed:
fix: Remove unnecessary ref specification in GitHub Actions workflow

Jovin-Ang committed:
fix: Update Java distribution in GitHub Actions workflow to Zulu

Jovin-Ang committed:
Merge pull request #3 from Jovin-Ang/branch-A-CI

Add GitHub Actions workflow for CI

Jovin-Ang committed:
fix: Standardize time format to use uppercase AM/PM in task details

Ensure consistent formatting of time in task details by standardizing
to use uppercase AM/PM. Previously, locale-specific differences caused
inconsistencies, with some systems using lowercase am/pm and others
using uppercase AM/PM. This led to test failures across different
environments.

Update the DateTimeFormatter to explicitly use the US locale, enforcing
uppercase AM/PM formatting. This change guarantees consistent behavior
regardless of the system's default locale.

Update tests to align with the standardized format.

Jovin-Ang committed:
fix: Handle storage loading failure by initializing an empty task list

Jovin-Ang committed:
feat: Add dialog box styling and layout adjustments

Jovin-Ang committed:
docs: Update README with detailed usage instructions for commands

Jovin-Ang committed:
docs: Update README for improved clarity and structure

Jovin-Ang committed:
fix: Add javafx-graphics dependency to build.gradle

Jovin-Ang committed:
fix: Trim task description to prevent empty input bypass

Ensure task descriptions are not just spaces by trimming input before
validation. Previously, isEmpty() returned false for space-only inputs.

Jovin-Ang committed:
docs: Update README to reflect latest version 1.0.2
